File: U1/EXAMEN/BanquitoServer_Soap_DotNet_GR01/BanquitoServer_Soap_DotNet_GR01/DataAccess/Repositories/MovimientoRepository.cs
using System;
using System.Collections.Generic;
using System.Linq;
using BanquitoServer_Soap_DotNet_GR01.Models;

namespace BanquitoServer_Soap_DotNet_GR01.DataAccess.Repositories
{
    /// <summary>
    /// Repositorio para operaciones de Movimiento
    /// </summary>
    public class MovimientoRepository
    {
        private readonly BanquitoDbContext _context;

        public MovimientoRepository()
        {
            _context = new BanquitoDbContext();
        }

        public MovimientoRepository(BanquitoDbContext context)
        {
            _context = context;
        }

        /// <summary>
        /// Verificar si el cliente tiene al menos un depósito en el último mes
        /// </summary>
        public bool TieneDepositoEnUltimoMes(string cedula)
        {
            var fechaInicio = DateTime.Now.AddMonths(-1);

            var tieneDeposito = (from m in _context.Movimientos
                                 join c in _context.Cuentas on m.CuentaId equals c.CuentaId
                                 join cl in _context.Clientes on c.ClienteId equals cl.ClienteId
                                 where cl.Cedula == cedula
                                    && m.TipoMovimiento == "DEPOSITO"
                                    && m.FechaMovimiento >= fechaInicio
                                 select m).Any();

            return tieneDeposito;
        }

        /// <summary>
        /// Obtener depósitos de un cliente en un período
        /// </summary>
        public List<Movimiento> FindDepositosByClienteAndPeriodo(string cedula, DateTime fechaInicio, DateTime fechaFin)
        {
            var depositos = (from m in _context.Movimientos
                            join c in _context.Cuentas on m.CuentaId equals c.CuentaId
                            join cl in _context.Clientes on c.ClienteId equals cl.ClienteId
                            where cl.Cedula == cedula
                               && m.TipoMovimiento == "DEPOSITO"
                               && m.FechaMovimiento >= fechaInicio
                               && m.FechaMovimiento <= fechaFin
                            select m).ToList();

            return depositos;
        }

        /// <summary>
        /// Obtener retiros de un cliente en un período
        /// </summary>
        public List<Movimiento> FindRetirosByClienteAndPeriodo(string cedula, DateTime fechaInicio, DateTime fechaFin)
        {
            var retiros = (from m in _context.Movimientos
                          join c in _context.Cuentas on m.CuentaId equals c.CuentaId
                          join cl in _context.Clientes on c.ClienteId equals cl.ClienteId
                          where cl.Cedula == cedula
                             && m.TipoMovimiento == "RETIRO"
                             && m.FechaMovimiento >= fechaInicio
                             && m.FechaMovimiento <= fechaFin
                          select m).ToList();

            return retiros;
        }

        /// <summary>
        /// Calcular promedio de depósitos en un período
        /// </summary>
        public decimal PromedioDepositosPorPeriodo(string cedula, DateTime fechaInicio, DateTime fechaFin)
        {
            var depositos = FindDepositosByClienteAndPeriodo(cedula, fechaInicio, fechaFin);

            if (depositos.Count == 0)
                return 0;

            return depositos.Average(d => d.Monto);
        }

        /// <summary>
        /// Calcular promedio de retiros en un período
        /// </summary>
        public decimal PromedioRetirosPorPeriodo(string cedula, DateTime fechaInicio, DateTime fechaFin)
        {
            var retiros = FindRetirosByClienteAndPeriodo(cedula, fechaInicio, fechaFin);

            if (retiros.Count == 0)
                return 0;

            return retiros.Average(r => r.Monto);
        }
    }
}

File: U2/PROYECTO/04. SERVIDOR/BanquitoServer_Restfull_Java_GR01/src/main/java/ec/edu/monster/Repositories/MovimientoRepository.java
package ec.edu.monster.Repositories;

import ec.edu.monster.Models.Movimiento;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

/**
 * Repositorio para operaciones de Movimiento
 */
@Repository
public interface MovimientoRepository extends JpaRepository<Movimiento, Long> {

    /**
     * Verificar si el cliente tiene al menos un depósito en el último mes
     */
    @Query("SELECT CASE WHEN COUNT(m) > 0 THEN true ELSE false END FROM Movimiento m " +
            "JOIN m.cuenta c " +
            "JOIN c.cliente cl " +
            "WHERE cl.cedula = :cedula " +
            "AND m.tipoMovimiento = 'DEPOSITO' " +
            "AND m.fechaMovimiento >= :fechaInicio")
    boolean tieneDepositoEnUltimoMes(@Param("cedula") String cedula,
            @Param("fechaInicio") LocalDateTime fechaInicio);

    /**
     * Obtener depósitos de un cliente en un período
     */
    @Query("SELECT m FROM Movimiento m " +
            "JOIN m.cuenta c " +
            "JOIN c.cliente cl " +
            "WHERE cl.cedula = :cedula " +
            "AND m.tipoMovimiento = 'DEPOSITO' " +
            "AND m.fechaMovimiento >= :fechaInicio " +
            "AND m.fechaMovimiento <= :fechaFin")
    List<Movimiento> findDepositosByClienteAndPeriodo(@Param("cedula") String cedula,
            @Param("fechaInicio") LocalDateTime fechaInicio,
            @Param("fechaFin") LocalDateTime fechaFin);

    /**
     * Obtener retiros de un cliente en un período
     */
    @Query("SELECT m FROM Movimiento m " +
            "JOIN m.cuenta c " +
            "JOIN c.cliente cl " +
            "WHERE cl.cedula = :cedula " +
            "AND m.tipoMovimiento = 'RETIRO' " +
            "AND m.fechaMovimiento >= :fechaInicio " +
            "AND m.fechaMovimiento <= :fechaFin")
    List<Movimiento> findRetirosByClienteAndPeriodo(@Param("cedula") String cedula,
            @Param("fechaInicio") LocalDateTime fechaInicio,
            @Param("fechaFin") LocalDateTime fechaFin);

    /**
     * Calcular promedio de depósitos en un período
     */
    @Query("SELECT COALESCE(AVG(m.monto), 0) FROM Movimiento m " +
            "JOIN m.cuenta c " +
            "JOIN c.cliente cl " +
            "WHERE cl.cedula = :cedula " +
            "AND m.tipoMovimiento = 'DEPOSITO' " +
            "AND m.fechaMovimiento >= :fechaInicio " +
            "AND m.fechaMovimiento <= :fechaFin")
    BigDecimal promedioDepositosPorPeriodo(@Param("cedula") String cedula,
            @Param("fechaInicio") LocalDateTime fechaInicio,
            @Param("fechaFin") LocalDateTime fechaFin);

    /**
     * Calcular promedio de retiros en un período
     */
    @Query("SELECT COALESCE(AVG(m.monto), 0) FROM Movimiento m " +
            "JOIN m.cuenta c " +
            "JOIN c.cliente cl " +
            "WHERE cl.cedula = :cedula " +
            "AND m.tipoMovimiento = 'RETIRO' " +
            "AND m.fechaMovimiento >= :fechaInicio " +
            "AND m.fechaMovimiento <= :fechaFin")
    BigDecimal promedioRetirosPorPeriodo(@Param("cedula") String cedula,
            @Param("fechaInicio") LocalDateTime fechaInicio,
            @Param("fechaFin") LocalDateTime fechaFin);
}
